README


This archive contains the distribution Pod-CYOA,
version 0.001:

  Pod-based Choose Your Own Adventure website generator (?!)

This software is copyright (c) 2010 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Pod-CYOA

0.001     2010-10-26 10:24:45 America/New_York
          first release


lib/Pod/CYOA.pm
use strict;
use warnings;
package Pod::CYOA;
BEGIN {
  $Pod::CYOA::VERSION = '0.001';
}
# ABSTRACT: Pod-based Choose Your Own Adventure website generator (?!)
1;


__END__
=pod

=head1 NAME

Pod::CYOA - Pod-based Choose Your Own Adventure website generator (?!)

=head1 VERSION

version 0.001

=head1 OVERVIEW

Pod::CYOA is a simple pair of libraries, L<Pod::CYOA::Transformer> and
L<Pod::CYOA::XHTML>, used to generate L<Choose Your Own
Adventure|http://en.wikipedia.org/wiki/Choose_Your_Own_Adventure> stories in
Pod.  It is not a robust, highly-extensible system, but it made available on
CPAN to make it easy to build the L<Dist::Zilla> tutorial, which is written in
Pod as a CYOA "story."

The L<Dist::Zilla tutorial|http://dzil.org/tutorial/start.html> itself includes
a L<complete sample program|http://dzil.org/tutorial/build-tutorial.html>
showing how Pod::CYOA is used to build the tutorial.  The L<dzil.org site
source code|http://github.com/rjbs/dzil.org> is also a useful place to look at
how to use Pod::CYOA.

=head1 STABILITY

Pod::CYOA makes no promises, other than that its latest version should work to
build the latest version of the Dist::Zilla tutorial.  It may grow more stable
over time if others are interested in using it.  Until that changes (and, thus,
this warning changes), you should be careful when upgrading Pod::CYOA, if you
have your own CYOA document sets.

=head1 AUTHOR

Ricardo SIGNES <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Pod/CYOA/XHTML.pm
use strict;
use warnings;
package Pod::CYOA::XHTML;
BEGIN {
  $Pod::CYOA::XHTML::VERSION = '0.001';
}
use Pod::Simple::XHTML 3.13;
use base 'Pod::Simple::XHTML';
# ABSTRACT: private Pod::Simple::XHTML subclass for CYOA


sub resolve_pod_page_link {
  my ($self, $to, $section) = @_;

  return $self->SUPER::resolve_pod_page_link($to, $section) if $to !~ s/^\@//;

  return "$to.html";
}

1;

__END__
=pod

=head1 NAME

Pod::CYOA::XHTML - private Pod::Simple::XHTML subclass for CYOA

=head1 VERSION

version 0.001

=head1 OVERVIEW

Pod::CYOA::HTML is a private class whose interface should not yet be relied
upon.

It transforms Pod to HTML, with one important change: links with targets that
begin with C<@> become links to relative html documents.  In other words,
C<< LE<lt>the start page|@startE<gt> >> becomes a link to C<start.html> rather
than the "start" section of the current page.

=head1 AUTHOR

Ricardo SIGNES <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Pod/CYOA/Transformer.pm
package Pod::CYOA::Transformer;
BEGIN {
  $Pod::CYOA::Transformer::VERSION = '0.001';
}
use Moose;
with 'Pod::Elemental::Transformer';
# ABSTRACT: transform 'cyoa' regions


use Pod::Elemental::Types qw(FormatName);

has format_name => (
  is  => 'ro',
  isa => FormatName,
  default => 'cyoa',
);

sub transform_node {
  my ($self, $node) = @_;

  for my $i (reverse(0 .. $#{ $node->children })) {
    my $para = $node->children->[ $i ];
    next unless $self->__is_xformable($para);

    my @replacements = $self->_expand_cyoa( $para );
    splice @{ $node->children }, $i, 1, @replacements;
  }
}

sub __is_xformable {
  my ($self, $para) = @_;

  return unless $para->isa('Pod::Elemental::Element::Pod5::Region')
         and $para->format_name eq $self->format_name;

  confess("CYOA regions must be non-pod (=begin " . $self->format_name . ")")
    if $para->is_pod;
  
  return 1;
}

sub _expand_cyoa {
  my ($self, $para) = @_;

  my ($data, @wtf) = @{ $para->children };
  confess "more than one child of a non-Pod region!" if @wtf;

  my @replacements;

  push @replacements, Pod::Elemental::Element::Pod5::Region->new({
    is_pod      => 0,
    format_name => 'html',
    content     => '',
    children    => [
      Pod::Elemental::Element::Pod5::Data->new({
        content => "<div class='cyoa'>",
      }),
    ],
  });

  push @replacements, Pod::Elemental::Element::Pod5::Command->new({
    command => 'over',
    content => 4,
  });

  my @lines = split /\n/, $data->as_pod_string;
  for my $line (@lines) {
    my ($link, $desc) = $line =~ /\A\?\s*([-a-z0-9]+)\s*\?\s*(.+)\z/;
    confess "do not understand CYOA line: $line" unless $link and $desc;

    push @replacements, Pod::Elemental::Element::Pod5::Command->new({
      command => 'item',
      content => '*',
    });

    push @replacements, Pod::Elemental::Element::Pod5::Ordinary->new({
      content => "If you'd like to L<$desc|\@$link>",
    });
  }

  push @replacements, Pod::Elemental::Element::Pod5::Command->new({
    command => 'back',
    content => '',
  });

  push @replacements, Pod::Elemental::Element::Pod5::Region->new({
    is_pod      => 0,
    format_name => 'html',
    content     => '',
    children    => [
      Pod::Elemental::Element::Pod5::Data->new({
        content => "</div>",
      }),
    ],
  });

  return @replacements;
}

1;

__END__
=pod

=head1 NAME

Pod::CYOA::Transformer - transform 'cyoa' regions

=head1 VERSION

version 0.001

=head1 OVERVIEW

Pod::CYOA::Transformer is a L<Pod::Elemental::Transformer> implementation.  It
looks for a region with the format name C<cyoa> and transforms it into a
C<=item>-list surrounded by C<html> regions.

A C<cyoa> region is written with pairs of C<?>-separated values representing
page links and descriptions.  For example:

  =for :cyoa
  ? pie-eating  ? eat a pie
  ? start       ? start over
  ? visit-lefty ? buy an "O"

...will become something like:

  =for html
  <div class='cyoa'>

  =over 4

  =item * If you'd like to L<eat a pie|@pie-eating>

  =item * If you'd like to L<start over|@start>

  =item * If you'd like to L<buy an "O"|@visit-lefty>

  =back

  =for html
  </div>

The C<@>-prefix on the link targets is expected to be handled by
L<Pod::CYOA::XHTML>.

=head1 AUTHOR

Ricardo SIGNES <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Pod-based Choose Your Own Adventure website generator (?!)',
  'AUTHOR' => 'Ricardo SIGNES <rjbs@cpan.org>',
  'BUILD_REQUIRES' => {
    'Test::More' => '0.96'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'Pod-CYOA',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Pod::CYOA',
  'PREREQ_PM' => {
    'Moose' => '0',
    'Pod::Elemental::Transformer' => '0',
    'Pod::Elemental::Types' => '0',
    'Pod::Simple::XHTML' => '3.13'
  },
  'VERSION' => '0.001',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



